### PR TITLE
Seminar 3 code refactoring

### PR DESCRIPTION
**Pull request by DenisIndenbom**
# Refactored code from Seminar 3 (#2)

Changes:
- Renamed enum `TableEnum` to `TableIndexes` ([here](https://github.com/D1ffic00lt/DSBA-Python-2024/pull/2#discussion_r1772801815))
- Artist, Url, Track and ets. transferred to the upper case ([here](https://github.com/D1ffic00lt/DSBA-Python-2024/pull/2#discussion_r1772804110))
- Added a check that a song with same id does not exist ([here](https://github.com/D1ffic00lt/DSBA-Python-2024/pull/2#discussion_r1772806551))
- Added a temp solution of this [problem](https://github.com/D1ffic00lt/DSBA-Python-2024/pull/2#discussion_r1772809752)
- Replaced `-> Song` to `-> Song | None` in `get_song` function ([here](https://github.com/D1ffic00lt/DSBA-Python-2024/pull/2#discussion_r1772812112))
- Other small changes

If u have questions or other requests in the review, then tell me about it.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,7 +3,7 @@
 **/__pycache__/
 *.py[cod]
 *$py.class
-
+.idea
```
Comment (D1ffic00lt) [inline .gitignore]: u are already have an .idea ignoring in the bottom 

```suggestion

```

File: .gitignore [hunk 2]
```diff
@@ -160,4 +160,4 @@ cython_debug/
 #  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
-#.idea/
\ No newline at end of file
+.idea/
```
Comment (D1ffic00lt) [inline .gitignore]: it will be better if u handle all directories 

```suggestion
**/.idea/
```

File: services/scripts.py [hunk 3]
```diff
@@ -1,61 +1,66 @@
...
+    def __parse_row(row: list[str]) -> Song:
+        row[TableIndexes.ID.value] = int(row[TableIndexes.ID.value])
 
-    def get_song(self, song_id: int) -> Song:
-        return self.db.get(song_id)
+        return Song(row)
 
+    def add_song(self, song: Song) -> None:
+        if song.id in self.db:
+            raise 'A song with this id already exists!'
```
Comment (D1ffic00lt) [inline services/scripts.py]: raising of an error without an error...

File: services/scripts.py [hunk 4]
```diff
@@ -1,61 +1,66 @@
 # Music recommendation application
 import csv
 
-from cfg import TableEnum
+from cfg import TableIndexes
 
 
 class Song:
     def __init__(self, row: list[str]) -> None:
-        self.id = int(row[TableEnum.ID.value])
```
Comment (D1ffic00lt) [inline services/scripts.py:9]: why..? as far as i know, we did it on purpose.. but i could be wrong

**Review by D1ffic00lt — COMMENTED** [review]
first of all, don't tag me in telegram, I have notifications of new pull requests, if you need a review, just mention me and @bani2016up on the pull request page
image

secondly, please, call commits by their content, please.. 

thirdly, below there are some suggestions for your code

File: services/scripts.py (hunk 4 above)
Comment (DenisIndenbom) [reply services/scripts.py:9]: Because I parse it in function `__parse_row`.

File: services/scripts.py (hunk 3 above)
Comment (DenisIndenbom) [reply services/scripts.py]: Oops

File: services/scripts.py (hunk 4 above)
Comment (D1ffic00lt) [reply services/scripts.py:9]: > Because I parse it in function `__parse_row`.

sry, my bad

File: services/scripts.py [hunk 5]
```diff
@@ -1,61 +1,66 @@
...
+    def __parse_row(row: list[str]) -> Song:
+        row[TableIndexes.ID.value] = int(row[TableIndexes.ID.value])
 
-    def get_song(self, song_id: int) -> Song:
-        return self.db.get(song_id)
+        return Song(row)
 
+    def add_song(self, song: Song) -> None:
+        if song.id in self.db:
+            raise Exception('A song with this id already exists!')
```
Comment (D1ffic00lt) [inline services/scripts.py]:
```suggestion
            raise ValueError('A song with this id already exists!')
```
I think it should be a `ValueError` (as a bad value, idk), maybe not, but it shouldn't be a simple `Exception`

**Review by D1ffic00lt — APPROVED** [review]
LGTM, but please fix the error class

**Comment by DenisIndenbom** [comment]
> LGTM, but please fix the error class 

Fixed

**Comment by DenisIndenbom** [comment]
Can we apply pull request without code review from @bani2016up ?  Or do we have to?

**Comment by bani2016up** [comment]
@DenisIndenbom, no, you can not, you need at least two approves.

File: cl_app.py [hunk 6]
```diff
@@ -1,28 +1,27 @@
 from services import get_shape, get_minimum_and_maximum, read_from_file, get_top_artists, Song
 
 if __name__ == "__main__":
```
Comment (bani2016up) [inline cl_app.py:3]: You need a main function here.

File: services/scripts.py [hunk 7]
```diff
@@ -1,61 +1,66 @@
...
             reader = csv.reader(file)
 
             for row in reader:
-                self.add_song(Song(row))
+                self.add_song(self.__parse_row(row))
 
-    def add_song(self, song: Song) -> None:
-        self.db[song.id] = song
+    @staticmethod
+    def __parse_row(row: list[str]) -> Song:
```
Comment (bani2016up) [inline services/scripts.py:27]: This is privet method, used sing _ sign, not __. More here: [https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name](click)

File: services/scripts.py [hunk 8]
```diff
@@ -65,3 +70,9 @@ def get_minimum_and_maximum(data: list[list[str]], index: int) -> tuple[float, f
 
 def get_shape(data: list[list[str]]) -> tuple[int, int]:
     return len(data), len(data[0])
+
+
+if __name__ == '__main__':
+    db = DataBase(filename='../data/Spotify_Youtube.csv')
+
+    print(db.get_song(1).name)
```
Comment (bani2016up) [inline services/scripts.py]: What if get_song returns None? This will cause an error: NoneType obj has no attr name

Comment (DenisIndenbom) [reply services/scripts.py]: So... I wrote that func can return none. I can write return of an empty object, but why?

File: cl_app.py (hunk 6 above)
Comment (DenisIndenbom) [reply cl_app.py:3]: Ok, I'll do it

File: services/scripts.py (hunk 7 above)
Comment (DenisIndenbom) [reply services/scripts.py:27]: Yeh, I know, but I want that this method not to be used in child classes.

Comment (D1ffic00lt) [reply services/scripts.py:27]: > This is privet method, used sing _ sign, not __. More here: [https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name](click)


i don’t thing it matters, in this case `__parse_row` should only be used in the `DataBase` class, so, `__…` can be here

Comment (DenisIndenbom) [reply services/scripts.py:27]: If it doesn't matter, then private or superprivate don't matter.

> in this case `__parse_row` should only be used in the `DataBase` class

So... In this case superprivate is more suitable...

File: services/scripts.py (hunk 8 above)
Comment (bani2016up) [reply services/scripts.py]: The problem is not in type hints. get method can return obj or None. At line 78 you say, form get_song take attr name. The problem is, that get_song can be both song and None. Does None has an name attr?

Comment (DenisIndenbom) [reply services/scripts.py]: It's just example.. I can remove it or fix it..

**Review by bani2016up — APPROVED** [review]

**Pull request merged**